vsock/src/thread_backend.rs

```diff
@@ -0,0 +1,306 @@
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+#![deny(missing_docs)]
+
+use super::{
+    rxops::*,
+    vhu_vsock::{
+        ConnMapKey, Error, Result, VSOCK_HOST_CID, VSOCK_OP_REQUEST, VSOCK_OP_RST,
+        VSOCK_TYPE_STREAM,
+    },
+    vhu_vsock_thread::VhostUserVsockThread,
+    vsock_conn::*,
+};
+use log::{info, warn};
+use std::{
+    collections::{HashMap, HashSet, VecDeque},
+    os::unix::{
+        net::UnixStream,
+        prelude::{AsRawFd, FromRawFd, RawFd},
+    },
+};
+use virtio_vsock::packet::VsockPacket;
+use vm_memory::bitmap::BitmapSlice;
+
+pub struct VsockThreadBackend {
+    /// Map of ConnMapKey objects indexed by raw file descriptors.
+    pub listener_map: HashMap<RawFd, ConnMapKey>,
+    /// Map of vsock connection objects indexed by ConnMapKey objects.
+    pub conn_map: HashMap<ConnMapKey, VsockConnection<UnixStream>>,
+    /// Queue of ConnMapKey objects indicating pending rx operations.
+    pub backend_rxq: VecDeque<ConnMapKey>,
+    /// Map of host-side unix streams indexed by raw file descriptors.
+    pub stream_map: HashMap<i32, UnixStream>,
+    /// Host side socket for listening to new connections from the host.
+    host_socket_path: String,
+    /// epoll for registering new host-side connections.
+    epoll_fd: i32,
+    /// Set of allocated local ports.
+    pub local_port_set: HashSet<u32>,
+}
+
+impl VsockThreadBackend {
+    /// New instance of VsockThreadBackend.
+    pub fn new(host_socket_path: String, epoll_fd: i32) -> Self {
+        Self {
+            listener_map: HashMap::new(),
+            conn_map: HashMap::new(),
+            backend_rxq: VecDeque::new(),
+            // Need this map to prevent connected stream from closing
+            // TODO: think of a better solution
+            stream_map: HashMap::new(),
+            host_socket_path,
+            epoll_fd,
+            local_port_set: HashSet::new(),
+        }
+    }
+
+    /// Checks if there are pending rx requests in the backend
+    /// rxq.
+    pub fn pending_rx(&self) -> bool {
+        !self.backend_rxq.is_empty()
+    }
+
+    /// Deliver a vsock packet to the guest vsock driver.
+    ///
+    /// Returns:
+    /// - `Ok(())` if the packet was successfully filled in
+    /// - `Err(Error::EmptyBackendRxQ) if there was no available data
+    pub(crate) fn recv_pkt<B: BitmapSlice>(&mut self, pkt: &mut VsockPacket<B>) -> Result<()> {
+        // Pop an event from the backend_rxq
+        let key = match self.backend_rxq.pop_front() {
+            Some(cmk) => cmk,
+            None => {
+                return Err(Error::EmptyBackendRxQ);
+            }
```

Comment:
What about:

        let key = match self.backend_rxq.pop_front().map_err(|_| Error::EmptyBackendRxQ)?;

Reply:
Yep, I'll change

Reply:
I just realized that the value returned by `pop_front()` is `std::option::Option` and `.map_err` is not available.
Is there any other nice way? otherwise I'll leave it like this.

Reply:
Maybe somthing like this ?

let key = self.backend_rxq.pop_front().ok_or(Error::EmptyBackendRxQ)?;